docs/TypeScript全面进阶指南/27.TSConfig 全解（上）：构建相关配置.md
---
title: 27.TSConfig 全解（上）：构建相关配置
date: 2023-10-09
---
在前面的内容中，我们已经学习了 TypeScript 在工程中的许多实践，包括类型声明、TypeScript 与 React、ESLint 的结合使用以及装饰器等。这些实践更像是上层建筑，默认是在一个已经基本配置完环境的 TypeScript 项目中进行的。这一节，我们深入下层基础，来了解 TypeScript 工程中最基础的一部分：TSConfig 配置。

为什么选择现在才讲配置呢？因为在前面的工程实践中，我们并不需要自己去修改 TSConfig，脚手架已经帮我们处理好了。有了实践经验，再来讲解讲解这些配置效果会更好。

为了避免罗列配置这种填鸭式教学，我将 TSConfig 分为三个大类：**构建相关**、**类型检查相关**以及**工程相关**。这其实也对应着我们的开发流程：使用工程能力进行项目开发，检查源码是否符合配置约束，然后才是输出产物。每一个大类又可以划分为几个小类，比如构建相关又可以分为**构建源码相关**与**构建产物相关**等等，我们会按照这些分类的方式进行聚合地讲解。

最后，正如我对这本小册的定位也包括工具书一样，当你在实际项目开发遗忘了某一项具体配置的作用，或者发现某一配置表现不符合预期，都可以回到这里来寻找答案。

这一节我们主要介绍构建相关的配置。

## 构建相关

### 构建源码相关

#### 特殊语法相关

##### experimentalDecorators 与 emitDecoratorMetadata

这两个选项都和装饰器有关，其中 experimentalDecorators 选项用于启用装饰器的 `@` 语法，而 emitDecoratorMetadata 配置则影响装饰器实际运行时的元数据相关逻辑，我们在装饰器一节中已经了解了此选项对实际编译代码的作用：

```javascript
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};

__decorate([
    Prop(),
    __metadata("design:type", String) // 来自于 emitDecoratorMetadata 配置，其它 __metadata 方法同
], Foo.prototype, "prop", void 0);

__decorate([
    Method(),
    __param(0, Param()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", void 0)
], Foo.prototype, "handler", null);

Foo = __decorate([
    Cls(),
    __param(0, Param()),
    __metadata("design:paramtypes", [String])
], Foo);
```

##### jsx、jsxFactory、jsxFragmentFactory 与 jsxImportSource

这部分配置主要涉及 jsx(tsx) 相关的语法特性。其中，jsx 配置将直接影响 JSX 组件的构建表现，常见的主要有 `react` （将 JSX 组件转换为对 `React.createElement` 调用，生成 `.js` 文件）、`preserve`（原样保留 JSX 组件，生成 `.jsx` 文件，你可以接着让其他的编译器进行处理）、`react-native` （类似于 preserve，但会生成 `.js` 文件）。

如果你希望使用特殊的 jsx 转换，也可以将其配置为 `react-jsx` / `react-jsxdev`，这样 JSX 组件会被转换为对 `__jsx` 方法的调用与生成 `.js` 文件，此方法来自于 `react/jsx-runtime`。

```jsx
// react
import React from 'react';
export const helloWorld = () => React.createElement("h1", null, "Hello world");

// preserve / react-native
import React from 'react';
export const helloWorld = () => <h1>Hello world</h1>;

// react-jsx
import { jsx as _jsx } from "react/jsx-runtime";
import React from 'react';
export const helloWorld = () => _jsx("h1", { children: "Hello world" });
 
// react-jsxdev
import { jsxDEV as _jsxDEV } from "react/jsx-dev-runtime";
const _jsxFileName = "/home/runner/work/TypeScript-Website/TypeScript-Website/index.tsx";
import React from 'react';
export const helloWorld = () => _jsxDEV("h1", { children: "Hello world" }, void 0, false, { fileName: _jsxFileName, lineNumber: 9, columnNumber: 32 }, this);
```

除了 jsx 以外，其它 jsx 相关配置使用较少，我们简单了解即可。

- jsxFactory，影响负责最终处理转换完毕 JSX 组件的方法，默认即为 `React.createElement`。如果你想使用 `preact.h` 作为处理方法，可以将其配置为 `h`。

- jsxFragmentFactory，类似 jsxFactory，只不过它影响的是 Fragment 组件（`<></>`）的提供方。jsxFactory 与 jsxFragmentFactory 均是 TS 4.1 版本以前用于实现自定义 JSX 转换的配置项，举例来说，当设置了 `"jsx": "react"`，此时将 jsxFragmentFactory 设置为 `Fragment` ，同时将 jsxFactory 设置为 `h`：

  ```jsx
  import { h, Fragment } from "preact";
  const HelloWorld = () => (
    <>
      <div>Hello</div>
    </>
  );
  
  // 转换为以下代码
  const preact_1 = require("preact");
  const HelloWorld = () => ((0, preact_1.h)(preact_1.Fragment, null,
      (0, preact_1.h)("div", null, "Hello")));
  ```

  为了简化自定义 JSX 转换的配置，4.1 版本以后 TS 支持使用 jsxImportSource 属性快速地调整。

- jsxImportSource，当你的 jsx 设置为 `react-jsx` / `react-jsxdev` 时，指定你的 `jsx-runtime` / `jsx-dev-runtime` 从何处导入。如设置为 `preact` 时，会从 `preact/jsx-runtime` 导入 `_jsx` 函数，用于进行 JSX 组件的转换。类似的，在另一个类 React 框架 Solid 中，也将此配置修改为了自己的实现： `"jsxImportSource": "solid-js"`。

这也是 React 17 中的变化之一，在 17 版本前后的构建后代码如下：

```js
// v17 前
import React from 'react';

function App() {
  return React.createElement('h1', null, 'Hello world');
}

// v17 后
import {jsx as _jsx} from 'react/jsx-runtime';

function App() {
  return _jsx('h1', { children: 'Hello world' });
}
```

这也是在 17 版本以后，不需要再确保代码中导入了 React 就能使用 JSX 的原因。

##### target 与 lib、noLib

target 配置决定了你的构建代码使用的语法，常用值包括 es5、es6、es2018、es2021、esnext（基于目前的 TypeScript 版本所支持的最新版本） 等等。某些来自于更高版本 ECMAScript 的语法，会在编译到更低版本时进行语法的降级，常见的如异步函数、箭头函数、bigint 数据类型等。

类似的，在 Babel 中也有 targets 的概念。但这里的 targets 通常指的是预期运行的浏览器，如 chrome 89，然后基于 browserlist 获取浏览器信息，基于 [caniuse](https://link.juejin.cn/?target=https%3A%2F%2Fcaniuse.com%2F) 或者 compat-table 获取各个浏览器版本支持的特性，最后再进行语法的降级。

如果没有特殊需要，推荐将 target 设置为 `"es2018"`，一个对常用语法支持较为全面的版本。

更改 target 配置也会同时影响你的 lib 配置默认值，而它决定了你是否能使用某些来自于更新版本的 ECMAScript 语法，以 replaceAll 为例，如果你直接在项目中使用：

```typescript
'linbudu'.replaceAll('d', 'dd');
```

此时，如果你的 lib 配置中不包含 `"es2021"` 或者 `"es2021.String"`，上面的代码就会给出一个错误提示：***属性“replaceAll”在类型“"linbudu"”上不存在。是否需要更改目标库? 请尝试将 “lib” 编译器选项更改为“es2021”或更高版本***。

正如我们在类型声明一节中了解的，TypeScript 会自动加载内置的 `lib.d.ts` 等声明文件，而加载哪些文件则和 lib 配置有关。当我们配置了 `"es2021"` 或者 `"es2021.String"`，replaceAll 方法对应的声明文件 `lib.es2021.string.d.ts` 就会被加载，然后我们的 String 类型上才有了 lib 方法。

除了高版本语法以外，lib 其实也和你的实际运行环境有关。比如，当你的代码仅在 Node 环境下运行时，你的 lib 中不应当包含 `"DOM"` 这个值。对应的，代码中无法使用 window 、document 等全局变量。

而 target 对 lib 的影响在于，当你的 target 为更高的版本时，它会自动地将这个版本新语法对应的 lib 声明加载进来，以上面的代码为例， target 为 `"es2021"` 时，你不需要添加 `"es2021"` 到 lib 中也能使用 ECMAScript2021 的新方法 replaceAll。这是因为既然你的编译产物都到这个版本了，那你当然可以直接使用这个方法啦。

如果你希望使用自己提供的 lib 声明定义，可以启用 noLib 配置，这样 TypeScript 将不会去加载内置的类型定义，但你需要为所有内置对象提供类型定义（String，Function，Object 等）才能进行编译。如果你的运行环境中存在大量的定制方法，甚至对原本的内置方法做了覆盖，就可以使用此配置来加载自己的类型声明。

最后，target 与 lib 配置会随着 TS 的版本更新而新增可用的值，如在 4.6 版本新增了 `es2022`这一选项，支持了 `Array.at()`、`Error Cause` 等新的语言特性。

### 构建解析相关

这部分配置主要控制源码解析，包括从何处开始收集要构建的文件，如何解析别名路径等等。

#### files、include 与 exclude

这三个选项决定了将被包括到本次编译的代码文件。使用 files 我们可以描述本次包含的所有文件，但不能使用 `src` 或者 `src/*` 这种方式，每个值都需要是完整的文件路径，适合在小型项目时使用：

```json
{
  "compilerOptions": {},
  "files": [
    "src/index.ts",
    "src/handler.ts"
  ]
}
```

如果你的文件数量较多，或者分散在各个文件夹，此时可以使用 include 和 exclude 进行配置，在这里可以传入文件夹或者 `src/*` 这样的 glob pattern，也可以传入完整的文件路径。

include 配置方式参考：

```json
{
  "include": ["src/**/*", "generated/*.ts", "internal/*"]
}
```

其中，`src/**/*` 表示匹配 src下所有的合法文件，而无视目录层级。而 `internal/*` 则只会匹配 internal 下的文件，不会匹配 `internal/utils/` 下的文件。这里的合法文件指的是，在不包括文件扩展名（`*.ts`）的情况下只会匹配 `.ts` / `.tsx` / `.d.ts` / `.js` / `.jsx` 文件（js 和 jsx 文件需要启用 allowJs 配置时才会被包括）。

由于我们会在 include 中大量使用 glob pattern 来一次性匹配许多文件，如果存在某些非预期的文件也符合这一匹配模式，比如 `src/handler.test.ts` `src/file-excluded/` 这样，此时专门为需要匹配的文件书写精确的匹配模式就太麻烦了。因此，我们可以使用 exclude 配置，来从被 include 匹配到的文件中再移除一部分，如：

```json
{
  "include": ["src/**/*", "generated/*.ts", "internal/*"],
  "exclude": ["src/file-excluded", "/**/*.test.ts", "/**/*.e2e.ts"]
}
```

需要注意的是，**exclude 只能剔除已经被 include 包含的文件**。

#### baseUrl

这一配置可以定义文件进行解析的根目录，它通常会是一个相对路径，然后配合 tsconfig.json 所在的路径来确定根目录的位置。

```text
project
├── out.ts
├── src
├──── core.ts
└── tsconfig.json
```

在这个结构下，如果配置为 `"baseUrl": "./"`，根目录就会被确定为 project。

你也可以通过这一配置，在导入语句中使用相对 baseUrl 的解析路径。如在上面根目录已经确定为 project，在 `out.ts` 中，你就可以直接使用基于根目录的绝对路径导入文件：

```typescript
import "src/core"; // TS 会自动解析到对应的文件，即 "./src/core.ts"
```

#### rootDir

rootDir 配置决定了项目文件的根目录，默认情况下它是项目内**包括**的所有 .ts 文件的最长公共路径，这里有几处需要注意：

- **包括**指的是 include 或 files 中包括的 `.ts` 文件，这些文件一般来说不会和 tsconfig.json 位于同一目录层级；
- 不包括 `.d.ts` 文件，因为声明文件可能会和 tsconfig.json 位于同一层级。

最长公共路径又是什么？简单地说，它就是某一个**包含了所有被包括的 `.ts` 文件的文件夹**，TypeScript 会找到这么一个文件夹，默认将其作为 rootDir。

```text
PROJECT
├── src
│   ├── index.ts
│   ├── app.ts
│   ├── utils
│   │   ├── helpers.ts
├── declare.d.ts
├── tsconfig.json
```

在这个例子中，rootDir 会被推断为 src。

```text
PROJECT
├── env
│   ├── env.dev.ts
│   ├── env.prod.ts
├── app
│   ├── index.ts
├── declare.d.ts
├── tsconfig.json
```

在这个例子中，rootDir 会被推断为 `.`，即 `tsconfig.json` 所在的目录。

构建产物的目录结构会受到这一配置的影响，假设 outDir 被配置为 `dist`，在上面的第一种情况下，最终的产物会被全部放置在 dist 目录下，保持它们在 `src`（也就是 rootDir） 内的目录结构：

```text
PROJECT
├── dist
│   ├── index.js
│   ├── index.d.ts
│   ├── app.js
│   ├── app.d.ts
│   ├── utils
│   │   ├── helpers.js
│   │   ├── helpers.d.ts
```

如果你将 rootDir 更改为推导得到的 rootDir 的父级目录，比如在这里把它更改到了项目根目录 `.`。此时 `src` 会被视为 rootDir 的一部分，因此最终构建目录结构中会多出 `src` 这一级：

```text
PROJECT
├── dist
├── ├──src
│      ├── index.js
│      ├── index.d.ts
│      ├── app.js
│      ├── app.d.ts
│      ├── utils
│      │   ├── helpers.js
│      │   ├── helpers.d.ts
```

需要注意的是，如果你显式指定 rootDir ，需要确保其包含了所有 **“被包括”** 的文件，因为 TypeScript 需要确保这所有的文件都被生成在 outDir 内。

```text
PROJECT
├── src
│   ├── index.ts
│   ├── app.ts
│   ├── utils
│   │   ├── helpers.ts
├── env.ts
├── tsconfig.json
```

在这个例子中，如果你指定 rootDir 为 `src` ，会导致 `env.ts` 被生成到 `<project>/env.js` 而非 `<project>/dist/env.js` 。

#### rootDirs

rootDirs 就是复数版本的 rootDir，它接收一组值，并且会将这些值均视为平级的根目录：

```json
{
  "compilerOptions": {
    "rootDirs": ["src/zh", "src/en", "src/jp"]
  }
}
PROJECT
├── src
│   ├── zh
│   │   ├── locale.ts
│   ├── en
│   │   ├── locale.ts
│   ├── jp
│   │   ├── locale.ts
│   ├── index.ts
├── tsconfig.json
```

使用 rootDirs，TypeScript 还是会隐式地推导 rootDir，此时它的值为 rootDirs 中所有文件夹最近的公共父文件夹，在这里即是 `src`。你肯定会想，那 rootDirs 还有什么用？实际上它主要用于实现**多个虚拟目录的合并解析**。还是以上面的例子为例，假设我们的目录结构是现在这样的：

```text
PROJECT
├── src
│   ├── locales
│   │   ├── zh.locale.ts
│   │   ├── en.locale.ts
│   │   ├── jp.locale.ts
│   ├── index.ts
│── generated
│   ├── messages
│   │   ├── main.mapper.ts
│   │   ├── info.mapper.ts
├── tsconfig.json
```

在这个目录结构中，`locales` 下存放我们定义的每个语言的对应翻译，`generated/messages` 则是通过扫描项目获得所有需要进行代码替换位置后生成的映射关系，我们在 `.locale.ts` 文件中会导入其中的 mapper 文件来生成对应的导出。

虽然现在 locale 文件和 mapper 文件被定义在不同的目录下，但在构建产物中它们实际上是位于同一层级的：

```text
│── dist
│   ├── zh.locale.js
│   ├── en.locale.js
│   ├── jp.locale.js
│   ├── main.mapper.js
│   ├── info.mapper.js
```

这也就意味着，我们应当是在 locale 文件中直接通过 `./main.mapper` 的路径来引用 mapper 文件的，而不是 `../../generated/messages/main.mapper.ts` 这样。

此时，我们就可以利用 rootDirs 配置来让 TS 将这两个相隔甚远的文件夹视为处于同一目录下：

```json
{
  "compilerOptions": {
    "rootDirs": ["src/locales", "generated/messages"]
  }
}
```

这一配置并不会影响实际的产物生成，它只会告诉 TS 将这两个模块视为同一层级下（类型定义层面）。

#### types 与 typeRoots

默认情况下，TypeScript 会加载 `node_modules/@types/` 下的所有声明文件，包括嵌套的 `../../node_modules/@types` 路径，这么做可以让你更方便地使用第三方库的类型。但如果你希望只加载实际使用的类型定义包，就可以通过 types 配置：

```json
{
  "compilerOptions": {
    "types": ["node", "jest", "react"]
  }
}
```

在这种情况下，只有 `@types/node`、`@types/jest` 以及 `@types/react` 会被加载。

即使其他 `@types/` 包没有被包含，它们也仍然能拥有完整的类型，但其中的全局声明（如 `process`，`expect`，`describe` 等全局变量）将不会被包含，同时也无法再享受到基于类型的提示。

如果你甚至希望改变加载 `@types/` 下文件的行为，可以使用 typeRoots 选项，其默认为 `@types`，即指定 `node_modules/@types` 下的所有文件（仍然包括嵌套的）。

```json
{
  "compilerOptions": {
    "typeRoots": ["./node_modules/@types", "./node_modules/@team-types", "./typings"],
    "types": ["react"],
    "skipLibCheck": true
  }
}
```

以上配置会尝试加载 `node_modules/@types/react` 以及 `./node_modules/@team-types/react` 、`./typings/react` 中的声明文件，注意我们需要使用**相对于 baseUrl 的相对路径**。

加载多个声明文件可能会导致内部的声明冲突，所以你可能会需要 skipLibCheck 配置来禁用掉对加载的类型声明的检查。

#### moduleResolution

这一配置指定了模块的解析策略，可以配置为 node 或者 classic ，其中 node 为默认值，而 classic 主要作向后兼容用，基本不推荐使用。

首先来看 node 解析模式，从名字也能看出来它其实就是与 node 一致的解析模式。假设我们有个 `src/index.js`，其中存在基于相对路径 `const foo = require("./foo")` 的导入，则会依次按照以下顺序解析：

- `/<root>/<project>/src/foo.js` 文件是否存在？

- ```
  /<root>/<project>/src/foo
  ```

   

  是否是一个文件夹？

  - 此文件夹内部是否包含 `package.json`，且其中使用 `main` 属性描述了这个文件夹的入口文件？
  - 假设 `main` 指向 `dist/index.js`，那这里会尝试寻找 `/<root>/<project>/src/foo/dist/index.js` 文件
  - 否则的话，说明这个文件不是一个模块或者没有定义模块入口，我们走默认的 `/foo/index.js` 。


.vuepress/config.js
module.exports = {
  title: "博客首页",
  description: "Vimalakirti blog",
  dest: "public",
  base: "/blogs/",
  head: [
    [
      "link",
      {
        rel: "icon",
        href: "/avatar.ico",
      },
    ],
    [
      "meta",
      {
        name: "viewport",
        content: "width=device-width,initial-scale=1,user-scalable=no",
      },
    ],
    // [
    //   'script',
    //   { "src": `https://unpkg.com/element-ui/lib/index.js` }
    // ],
    // [
    //   "link",
    //   {
    //     "rel": "stylesheet",
    //     "href": "https://unpkg.com/element-ui/lib/theme-chalk/index.css"
    //   }
    // ],
  ],
  plugins: {
    "@vuepress/pwa": {
      serviceWorker: true,
      updatePopup: {
        message: "新的风暴再次出现！",
        buttonText: "刷新",
      },
    },
    "demo-container": true,
    // "@vuepress/back-to-top": true,
    "@vuepress/active-header-links": {
      sidebarLinkSelector: ".sidebar-link",
      headerAnchorSelector: ".header-anchor",
    },
    "@vssue/vuepress-plugin-vssue": {
      // 设置 `platform` 而不是 `api`
      platform: "github-v4",
      // 其他的 Vssue 配置
      owner: "Vimalate",
      repo: "Vimalate",
      clientId: "7fabb016743a2a816a75",
      clientSecret: "b07104f78c7cb82b05c2a816645a6165846c0a93",
      autoCreateIssue: true,
    },
    "cursor-effects": {
      size: 2, // size of the particle, default: 2
      shape: "star", // ['star' | 'circle'], // shape of the particle, default: 'star'
      zIndex: 999999999, // z-index property of the canvas, default: 999999999
    },
    // "@vuepress-reco/vuepress-plugin-kan-ban-niang": {
    //   theme: [
    //     "blackCat",
    //     "whiteCat",
    //     "haru1",
    //     "haru2",
    //     "haruto",
    //     "koharu",
    //     "izumi",
    //     "shizuku",
    //     "wanko",
    //     "miku",
    //     "z16",
    //   ],
    // },
    // "vuepress-plugin-nuggets-style-copy": true,
    "one-click-copy": {
      copySelector: [
        'div[class*="language-"] pre',
        'div[class*="aside-code"] aside',
      ], // String or Array
      copyMessage: "复制成功！", // default is 'Copy successfully and then paste it for use.'
      showInMobile: false, // whether to display on the mobile side, default: false.
    },
    "vuepress-plugin-helper-live2d": {
      // 是否开启控制台日志打印(default: false)
      log: false,
      live2d: {
        // 是否启用(关闭请设置为false)(default: true)
        enable: true,
        // 模型名称(default: hibiki)>>>取值请参考：
        // https://github.com/JoeyBling/hexo-theme-yilia-plus/wiki/live2d%E6%A8%A1%E5%9E%8B%E5%8C%85%E5%B1%95%E7%A4%BA
        model: "wanko",
        display: {
          position: "right", // 显示位置：left/right(default: 'right')
          width: 135, // 模型的长度(default: 135)
          height: 300, // 模型的高度(default: 300)
          hOffset: 65, //  水平偏移(default: 65)
          vOffset: 0, //  垂直偏移(default: 0)
        },
        mobile: {
          show: true, // 是否在移动设备上显示(default: false)
        },
        react: {
          opacity: 0.8, // 模型透明度(default: 0.8)
        },
      },
    },
    "dynamic-title": {
      showText: "(/≧▽≦/)咦！又好了！",
      hideText: "(●—●)喔哟，崩溃啦！",
      recoverTime: 2000,
    },
    "ribbon-animation": {
      size: 90, // 默认数据
      opacity: 0.3, //  透明度
      zIndex: -1, //  层级
      opt: {
        // 色带HSL饱和度
        colorSaturation: "80%",
        // 色带HSL亮度量
        colorBrightness: "60%",
        // 带状颜色不透明度
        colorAlpha: 0.65,
        // 在HSL颜色空间中循环显示颜色的速度有多快
        colorCycleSpeed: 6,
        // 从哪一侧开始Y轴 (top|min, middle|center, bottom|max, random)
        verticalPosition: "center",
        // 到达屏幕另一侧的速度有多快
        horizontalSpeed: 200,
        // 在任何给定时间，屏幕上会保留多少条带
        ribbonCount: 2,
        // 添加笔划以及色带填充颜色
        strokeSize: 0,
        // 通过页面滚动上的因子垂直移动色带
        parallaxAmount: -0.5,
        // 随着时间的推移，为每个功能区添加动画效果
        animateSections: true,
      },
      ribbonShow: false, //  点击彩带  true显示  false为不显示
      ribbonAnimationShow: true, // 滑动彩带
    },
  },
  theme: "reco",
  themeConfig: {
    nav: [
      {
        text: "首页",
        link: "/",
        icon: "reco-home",
      },
      {
        text: "时间线",
        link: "/timeline/",
        icon: "reco-date",
      },
      {
        text: "文档",
        icon: "reco-message",
        items: [
          // {
          //   "text": "vuepress-reco",
          //   "link": "/docs/theme-reco/"
          // },
          {
            text: "HTML-Library",
            link: "/docs/HTML-Library/",
          },
          {
            text: "CSS-Library",
            link: "/docs/CSS-Library/",
          },
          {
            text: "Javascript-Library",
            link: "/docs/Javascript-Library/",
          },
          {
            text: "Vue-Library",
            link: "/docs/Vue-Library/",
          },
          {
            text: "React-Library",
            link: "/docs/React-Library/",
          },
          {
            text: "Other-Library",
            link: "/docs/Other-Library/",
          },
          {
            text: "juejin",
            link: "/docs/juejin/",
          },
          {
            text: "web-interview",
            link: "/docs/interview/",
          },
          {
            text: "前端性能优化与实践",
            link: "/docs/前端性能优化与实践/",
          },
          {
            text: "使用webpack定制前端开发环境",
            link: "/docs/使用webpack定制前端开发环境/",
          },
          {
            text: "TypeScript全面进阶指南",
            link: "/docs/TypeScript全面进阶指南/",
          },
        ],
      },
      {
        text: "Contact",
        icon: "reco-message",
        items: [
          {
            text: "GitHub",
            link: "https://github.com/Vimalate",
            icon: "reco-github",
          },
        ],
      },
    ],
    sidebar: {
      // "/docs/theme-reco/": [
      //   "",
      //   "theme",
      //   "plugin",
      //   "api"
      // ],
      "/docs/HTML-Library/": [
        // "",
        "HTML-Study",
        "HTML技巧",
      ],
      "/docs/CSS-Library/": [
        "",
        "那些奇妙的 CSS ",
        "我都知道的CSS",
        "CSS 技巧",
        "别再问我布局了喔",
      ],
      "/docs/Javascript-Library/": [
        "",
        "常见对象方法",
        "常见手写代码",
        "常见数组方法",
        "常见字符串方法",
        "字符串技巧",
        "解构赋值",
        "可选链式操作符",
        "空值合并运算符",
        "了解Event Loop么",
        "聊聊原型和原型链",
        "我都知道的JS",
        "ES6的Set ，WeakSet，Map和WeakMap",
        "for of 和 for in 及forEach",
        "Javascript 面试题",
        "js存储和垃圾回收",
        "js继承",
        "js跳出循环的方法区别break,continue,return",
        "实现一个 new 操作符",
        "reduce 的妙用",
      ],
      "/docs/Vue-Library/": [
        "",
        "$ref和$el区别",
        "常用mixin",
        "项目添加svgIcon",
        "修改Vue打包后文件的接口地址配置",
        "虚拟列表",
        "自定义指令",
        "axios封装",
        "Element-ui 全局默认属性配置",
        "evnetbus-事件总线",
        "router-view 的 key 属性",
        "vue 整理面试题",
        "Vue的provide、inject",
        "vue响应原理",
        "vue项目路由权限配置",
        "Vue项目启动时间太慢",
        "vue项目问题总结",
        "vue中的$attrs和$listeners",
        "关于组件二次封装的又一步思考",
      ],
      "/docs/React-Library/": ["", "react-study"],
      "/docs/Other-Library/": [
        "",
        "常见手写",
        "错误监控",
        "登录鉴权方式",
        "工具函数",
        "利器推荐",
        "面试总结",
        "设计模式",
        "模板方法模式，父类封装流程，子类具体实现",
        "那些面试的手写",
        "那些年，我们忘了的正则 ×",
        "前端模块化：CommonJS,AMD,CMD,ES6",
        "前端协作规范",
        "去掉那些if else",
        "使用signature_pad实现数字签名",
        "算法初探",
        "图片上传压缩",
        "verdaccio搭建npm私有仓库",
        "网络安全",
        "微前端qiankun实践",
        "手写 mini qiankun",
        "移动端常见兼容问题",
        "移动端适配",
        "原生内嵌webview",
        "axios token失效取消后续请求",
        "axios.all和axios.spread",
        "nvm问题处理",
        "excel指定行导入",
        "HTTP",
        "http协议",
        "js函数工具库",
        "markdown语法",
        "nginx配置",
        "Oauth2.0",
        "print 打印",
        "session、cookie、Token和JWT",
        "Vite初识",
        "一些工作中值得收藏的网站",
        "pnpm、vue3搭建组件库",
        "WebRtc初探",
        "Echarts常用配置",
      ],
      "/docs/juejin/": [
        "",
        "20个js工具函数助力高效开发",
        "当我们二次封装时我们在封装什么",
        "关于首屏优化，我做了哪些",
        "使用VuePress搭建Element的组件库文档",
        "Typescript学习指南",
        "vue3正式发布两年后，我才开始学（2w字总结）",
        "微前端实践",
        "我迟来的 2022 年终总结：前已无通路，后不见归途",
        "如何使用 pnpm+vue3+vite 搭建组件库并发布到私有仓库",
        "常见手写代码",
        "vue3 里使用jsx",
      ],
      "/docs/interview/": [
        "",
        "2023-web-interview",
        "简历准备",
        "面试题整理",
        "计算机与网络安全",
        "工程化",
        "html",
        "js",
        "css",
        "vue",
      ],
      "/docs/前端性能优化与实践/": [
        "",
        "开篇：知识体系与小册格局",
        "存储篇 1：浏览器缓存机制介绍与缓存策略剖析",
        "存储篇 2：本地存储——从 Cookie 到 Web Storage、IndexDB",
        "前方的路：希望以此为你的起点",
        "网络篇 1：webpack 性能调优与 Gzip 原理",
        "网络篇 2：图片优化——质量与性能的博弈",
        "性能监测篇：Performance、LightHouse 与性能 API",
        "渲染篇 1：服务端渲染的探索与实践",
        "渲染篇 2：知己知彼——解锁浏览器背后的运行机制",
        "渲染篇 3：对症下药——DOM 优化原理与基本实践",
        "渲染篇 4：千方百计——Event Loop 与异步更新策略",
        "应用篇 1：优化首屏体验——Lazy-Load 初探",
        "应用篇 2：事件的节流（throttle）与防抖（debounce）",
        "彩蛋篇：CDN 的缓存与回源机制解析",
      ],
      "/docs/TypeScript全面进阶指南/": [
        "",
        "1.开篇：用正确的方式学习 TypeScript",
        "2.工欲善其事：打造最舒适的 TypeScript 开发环境",
        "3.进入类型的世界：理解原始类型与对象类型",
        "4.掌握字面量类型与枚举，让你的类型再精确一些",
        "5.函数与 Class 中的类型：详解函数重载与面向对象",
        "6.探秘内置类型：any、unknown、never 与类型断言",
        "7.类型编程好帮手：TypeScript 类型工具（上）",
        "8.类型编程好帮手：TypeScript 类型工具（下）",
        "9.类型编程基石：TypeScript 中无处不在的泛型",
        "10.结构化类型系统：类型兼容性判断的幕后",
        "11.类型系统层级：从 Top Type 到 Bottom Type",
        "12.类型里的逻辑运算：条件类型与 infer",
        "13.内置工具类型基础：别再妖魔化工具类型了！",
        "14.反方向类型推导：用好上下文相关类型",
        "15.数类型：协变与逆变的比较",
        "16.了解类型编程与类型体操的意义，找到平衡点",
        "17.内置工具类型进阶：类型编程进阶",
        "18.基础类型新成员：模板字符串类型入门",
        "19.类型编程新范式：模板字符串工具类型进阶",
        "20.工程层面的类型能力：类型声明、类型指令与命名空间",
        "21.在 React 中愉快地使用 TypeScript：内置类型与泛型坑位",
        "22.让 ESLint 来约束你的 TypeScript 代码：配置与规则集介绍",
        "23.全链路 TypeScript 工具库，找到适合你的工具",
        "24.说说 TypeScript 和 ECMAScript 之间那些事儿",
        "25.装饰器与反射元数据：了解装饰器基本原理与应用",
        "26.控制反转与依赖注入：基于装饰器的依赖注入实现",
        "27.TSConfig 全解（上）：构建相关配置",
        "28.TSConfig 全解（下）：检查相关、工程相关配置",
        "29.基于 Prisma + NestJs 的 Node API ：前置知识储备",
        "30.基于 Prisma + NestJs 的 Node API ：项目开发与基于 Heroku 部署",
        "31.玩转 TypeScript AST：AST Checker 与 CodeMod",
        "32.感谢相伴：是结束，也是开始",
        "33.漫谈篇：面试中的 TypeScript",
      ],
    },
    type: "blog",
    blogConfig: {
      category: {
        location: 2,
        text: "分类",
      },
      tag: {
        location: 3,
        text: "标签",
      },
    },
    // "friendLink": [
    //   {
    //     "title": "午后南杂",
    //     "desc": "Enjoy when you can, and endure when you must.",
    //     "email": "1156743527@qq.com",
    //     "link": "https://www.recoluan.com"
    //   },
    //   {
    //     "title": "vuepress-theme-reco",
    //     "desc": "A simple and beautiful vuepress Blog & Doc theme.",
    //     "avatar": "https://vuepress-theme-reco.recoluan.com/icon_vuepress_reco.png",
    //     "link": "https://vuepress-theme-reco.recoluan.com"
    //   }
    // ],
    logo: "/avatar.jpg",
    search: true,
    searchMaxSuggestions: 10,
    lastUpdated: "Last Updated",
    author: "呛再首",
    authorAvatar: "/avatar.jpg",
    record: "回到顶部",
    startYear: "2019",
  },
  markdown: {
    lineNumbers: true,
  },
  chainWebpack: (config) => {
    config.resolve.alias.set("core-js/library/fn", "core-js/features");
  },
};
